### PR TITLE
Modify to check if alpha is in host memory.

### PR DESCRIPTION
**Pull request by geotseng-amd**
# Summary

In this PR, we modified the runtime verifications to check if **Alpha** is in host memory.

Due to a modification from PR#956, the CPU may try to access device memory through **Alpha**, which violates the BIOS restrictions.

Assigned Ticket: https://ontrack-internal.amd.com/browse/SWDEV-496511
Caused Ticket: https://ontrack-internal.amd.com/browse/SWDEV-482839
Caused PR: https://github.com/ROCm/hipBLASLt/pull/956/

# Modifications
## In hipBLASLt API
HIPBLASLT_MATMUL_DESC_POINTER_MODE has three modes as the following:
```
typedef enum {
    HIPBLASLT_POINTER_MODE_HOST = 0,                          /** targets host memory */
    HIPBLASLT_POINTER_MODE_DEVICE = 1,                        /** targets device memory */
    HIPBLASLT_POINTER_MODE_ALPHA_DEVICE_VECTOR_BETA_HOST = 4, /** alpha pointer targets a device memory vector of length equal to the number of rows of matrix D, and beta is a single value in host memory. */
} hipblasLtPointerMode_t;
```
Only the HIPBLASLT_POINTER_MODE_HOST will ensure alpha is in host memory, so we modified **validateMatmulArgs** to check **pointermode**.

## In hipBLASLtExt API
Since this ScaleAlphaVec is passed as a parameter through the API, we can simply push a nullptr as **pointermode**.

### CODE REVIEW

**Comment by geotseng-amd** [comment]
All the CI local tests are passed for PR#1326

[0;32m[----------] [mGlobal test environment tear-down
[0;32m[==========] [m48760 tests from 13 test suites ran. (1755940 ms total)
[0;32m[  PASSED  ] [m48760 tests.
hipBLASLt version: 1100

command line: ./build/release/clients/staging/hipblaslt-test --gtest_output=xml --gtest_color=yes --gtest_repeat=1

File: library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp [hunk 1]
```diff
@@ -163,7 +163,8 @@ inline rocblaslt_status validateMatmulArgs(int64_t                m,
                                            int64_t                batch_stride_a = 0,
                                            int64_t                batch_stride_b = 0,
                                            int64_t                batch_stride_c = 0,
-                                           int64_t                batch_stride_d = 0)
+                                           int64_t                batch_stride_d = 0, 
+                                           const void*            pointermode    = nullptr) // add pointermode to check if alpha is in device memory.
```
Comment (jichangjichang) [inline library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp:167]: pointer mode is enum but not void*

File: library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp [hunk 2]
```diff
@@ -243,8 +244,8 @@ inline rocblaslt_status validateMatmulArgs(int64_t                m,
         return rocblaslt_status_invalid_pointer;
 
     // pointers must be valid
-    // Update for the valid case: (alpha=0 && (A=NULL || B=NULL))
-    if(n && ((k && (!alpha || ((*((float*)alpha)) && (!a || !b)))) || !c || !d))
+    // Update for the valid case: ((alpha_in_host && alpha=0) && (A=NULL || B=NULL))
```
Comment (jichangjichang) [inline library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp:247]: separate the checking, too many condition in one "if"

File: library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp [hunk 3]
```diff
@@ -831,7 +832,8 @@ rocblaslt_status rocblaslt_gemm_create_cpp_impl_2(const rocblaslt_handle handle,
                                      batch_stride_a,
                                      batch_stride_b,
                                      batch_stride_c,
-                                     batch_stride_d);
+                                     batch_stride_d,
+                                     nullptr); // add pointermode(nullptr) to check if alpha is in device memory.
```
Comment (jichangjichang) [inline library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp]: why did you just set nullptr for checking?

Comment (geotseng-amd) [reply library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp]: 1. In hipBLASLt API, pointermode==nullptr means alpha is in host memory
2. In hipBLASLtExt API, there is no pointer mode, and scalealphavec is passed through another parameter. so we can simply pass a nullptr to check.

File: library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp [hunk 4]
```diff
@@ -243,9 +243,10 @@ inline rocblaslt_status validateMatmulArgs(int64_t                m,
     if(!beta)
         return rocblaslt_status_invalid_pointer;
 
-    // pointers must be valid
     // Update for the valid case: ((alpha_in_host && alpha=0) && (A=NULL || B=NULL))
-    if(n && ((k && (!alpha || ((pointermode || (*((float*)alpha))) && (!a || !b)))) || !c || !d))
+    bool alpha_A_B_violation = ((pointermode || (*((float*)alpha))) && (!a || !b));
```
Comment (jichangjichang) [inline library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp]: you try to access alpha's data before checking if alpha is nullptr.

**Comment by geotseng-amd** [comment]
This PR has been verified correctness in the CQE machine.

[----------] Global test environment tear-down
[==========] 12397 tests from 13 test suites ran. (1879985 ms total)
[  PASSED  ] 12397 tests.
hipBLASLt version: 1100

command line: ./hipblaslt-test 

INFO - [CMD OUTPUT ENDS] 
INFO - ********************************************************************************
INFO - 
+----------------------+
|   Execution Result   |
+----------------------+
| total execution time |
+----------------------+
|  1883.0161821842194  |
+----------------------+
INFO - max previous execution time (For timeout)=0.8999569416046143 , Taken from DB with hostname=cqe-nv31-1-ub-hy-stg-nv31-04
INFO - 
+------------------------------------------------------------------+
|                        System Information                        |
+-----------------------+------------------------------------------+
|       Component       |               Information                |
+-----------------------+------------------------------------------+
|      rocm version     |                   2144                   |
|           OS          |                  Ubuntu                  |
|       OS Version      |                  22.04                   |
|     OS Pretty Name    |            Ubuntu 22.04.3 LTS            |
|     vbios version     |             113-D7070100-100             |
|                       |                                          |
|     sbios version     |                   4001                   |
|   GPU marketing name  |           AMD Radeon PRO W7900           |
|        GPU arch       |                 gfx1100                  |
|     Number of GPUs    |                    1                     |
|     GPU partition     |          Compute: N/A, Mem: N/A          |
|     CPU model name    | Intel(R) Core(TM) i7-6800K CPU @ 3.40GHz |
|        hostname       |       cqe-nv31-1-ub-hy-stg-nv31-04       |
|  SMC Firmware version |               00.78.128.00               |
|                       |                                          |
| SDMA Firmware version |                    24                    |
|                       |                                          |
|      XGMI setup?      |                    No                    |
+-----------------------+------------------------------------------+
INFO - 
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|                                                                                  Summary Table                                                                                  |
+------------------+-----------------+----------------------------+----------------------------+-----------------------------+-------------------------------------+--------------+
| No of TestSuites | No of TestCases | Number of Passed testcases | Number of Failed testcases | Number of Errored testcases | Number of Skipped/Not run testcases | Final Result |
+------------------+-----------------+----------------------------+----------------------------+-----------------------------+-------------------------------------+--------------+
|        13        |      12397      |           12397            |             0              |              0              |                  0                  |     PASS     |
+------------------+-----------------+----------------------------+----------------------------+-----------------------------+-------------------------------------+--------------+
[2144_nv31_hipblaslt.txt](https://github.com/user-attachments/files/17772490/2144_nv31_hipblaslt.txt)

File: library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp [hunk 5]
```diff
@@ -658,9 +659,9 @@ rocblaslt_status rocblaslt_matmul(rocblaslt_handle             handle,
     }
 
     // Check if pointer is valid
-    // Update for the valid case: (alpha=0 && (A=NULL || B=NULL))
+    // Update for the valid case: ((alpha_in_host && alpha=0) && (A=NULL || B=NULL))
     if(alpha == nullptr || beta == nullptr || C == nullptr || D == nullptr
-       || ((*((float*)alpha)) && (A == nullptr || B == nullptr)))
+       || ((matmul_descr->pointermode || (*((float*)alpha))) && (A == nullptr || B == nullptr)))
```
Comment (jichangjichang) [inline library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp]: separate the checking, too many condition in one "if"

File: library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp [hunk 6]
```diff
@@ -140,30 +140,31 @@ inline rocblaslt_status validateMatmulDescrArgs(rocblaslt_handle       handle,
...
+                                           hipblasOperation_t            opB,
+                                           int                           num_batches_a  = 1,
+                                           int                           num_batches_b  = 1,
+                                           int                           num_batches_c  = 1,
+                                           int                           num_batches_d  = 1,
+                                           int64_t                       batch_stride_a = 0,
+                                           int64_t                       batch_stride_b = 0,
+                                           int64_t                       batch_stride_c = 0,
+                                           int64_t                       batch_stride_d = 0, 
+                                           const rocblaslt_pointer_mode& pointermode    = rocblaslt_pointer_mode_host) // add pointermode to check if alpha is in device memory.
```
Comment (jichangjichang) [inline library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp]: did you run clang format?
remove comment
 // add pointermode to check if alpha is in device memory.

File: library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp [hunk 7]
```diff
@@ -405,7 +407,8 @@ inline rocblaslt_status rocblaslt_matmul_valid_args(const rocblaslt_matmul_desc
                                      batch_stride_a,
                                      batch_stride_b,
                                      batch_stride_c,
-                                     batch_stride_d);
+                                     batch_stride_d,
+                                     matmul_descr->pointermode); // add pointermode to check if alpha is in device memory.
```
Comment (jichangjichang) [inline library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp:413]: remove comment

File: library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp [hunk 8]
```diff
@@ -455,7 +455,8 @@ rocblaslt_status
                                             batch_stride_a,
                                             batch_stride_b,
                                             batch_stride_c,
-                                            batch_stride_d);
+                                            batch_stride_d,
+                                            matmul_descr[i]->pointermode); // add pointermode to check if alpha is in device memory.
```
Comment (jichangjichang) [inline library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp]: remove comment

**Review by jichangjichang — APPROVED** [review]

**Review by cmingch — APPROVED** [review]

**Pull request merged**